File: backend/src/services/googlePlaces.ts
import axios from 'axios';
import dotenv from 'dotenv';

dotenv.config();

const API_KEY = process.env.GOOGLE_MAPS_API_KEY;
const BASE_URL = 'https://maps.googleapis.com/maps/api/place';

export interface PlaceBusiness {
  place_id: string;
  name: string;
  vicinity: string;
  formatted_address?: string;
  geometry: { location: { lat: number; lng: number } };
  rating?: number;
  user_ratings_total?: number;
  price_level?: number;
  types?: string[];
  photos?: Array<{ photo_reference: string; width: number; height: number }>;
  opening_hours?: { open_now?: boolean };
  business_status?: string;
}

export interface PlaceDetail {
  place_id: string;
  name: string;
  formatted_address: string;
  formatted_phone_number?: string;
  website?: string;
  rating?: number;
  user_ratings_total?: number;
  price_level?: number;
  types?: string[];
  geometry: { location: { lat: number; lng: number } };
  opening_hours?: {
    open_now?: boolean;
    weekday_text?: string[];
    periods?: Array<{
      open: { day: number; time: string };
      close?: { day: number; time: string };
    }>;
  };
  photos?: Array<{ photo_reference: string; width: number; height: number }>;
  reviews?: Array<{
    author_name: string;
    rating: number;
    text: string;
    time: number;
    profile_photo_url?: string;
  }>;
  address_components?: Array<{
    long_name: string;
    short_name: string;
    types: string[];
  }>;
}

const mapTypeToCategory = (types: string[]): string => {
  const typeMap: Record<string, string> = {
    restaurant: 'Food & Drink',
    food: 'Food & Drink',
    meal_delivery: 'Food & Drink',
    meal_takeaway: 'Food & Drink',
    bakery: 'Food & Drink',
    bar: 'Food & Drink',
    cafe: 'Coffee & Tea',
    store: 'Shopping',
    shopping_mall: 'Shopping',
    clothing_store: 'Shopping',
    grocery_or_supermarket: 'Shopping',
    beauty_salon: 'Health & Beauty',
    hair_care: 'Health & Beauty',
    spa: 'Health & Beauty',
    gym: 'Fitness',
    car_repair: 'Automotive',
    car_dealer: 'Automotive',
    car_wash: 'Automotive',
    movie_theater: 'Entertainment',
    amusement_park: 'Entertainment',
    night_club: 'Entertainment',
    art_gallery: 'Arts & Culture',
    museum: 'Arts & Culture',
    plumber: 'Services',
    electrician: 'Services',
    lawyer: 'Services',
    accounting: 'Services',
    dentist: 'Services',
    doctor: 'Services',
    hospital: 'Services',
  };

  for (const type of types) {
    if (typeMap[type]) return typeMap[type];
  }
  return 'Other';
};

export const searchNearby = async (
  latitude: number,
  longitude: number,
  radius: number = 5000,
  keyword?: string,
  type?: string,
): Promise<PlaceBusiness[]> => {
  if (!API_KEY) {
    console.warn('Google Maps API key not set');
    return [];
  }

  try {
    const params: Record<string, any> = {
      location: `${latitude},${longitude}`,
      radius: Math.min(radius, 50000),
      type: type || 'establishment',
      key: API_KEY,
    };
    if (keyword) params.keyword = keyword;

    const response = await axios.get(`${BASE_URL}/nearbysearch/json`, { params });

    if (response.data.status !== 'OK' && response.data.status !== 'ZERO_RESULTS') {
      console.error('Places API error:', response.data.status, response.data.error_message);
      return [];
    }

    return response.data.results || [];
  } catch (error) {
    console.error('Google Places nearby search error:', error);
    return [];
  }
};

export const getPlaceDetails = async (placeId: string): Promise<PlaceDetail | null> => {
  if (!API_KEY) return null;

  try {
    const response = await axios.get(`${BASE_URL}/details/json`, {
      params: {
        place_id: placeId,
        fields: 'place_id,name,formatted_address,formatted_phone_number,website,rating,user_ratings_total,price_level,types,geometry,opening_hours,photos,reviews,address_components',
        key: API_KEY,
      },
    });

    if (response.data.status !== 'OK') {
      console.error('Place details error:', response.data.status);
      return null;
    }

    return response.data.result;
  } catch (error) {
    console.error('Google Places detail error:', error);
    return null;
  }
};

export const getPhotoUrl = (photoReference: string, maxWidth: number = 400): string => {
  return `${BASE_URL}/photo?maxwidth=${maxWidth}&photo_reference=${photoReference}&key=${API_KEY}`;
};

export const formatPlaceAsBusiness = (place: PlaceBusiness) => {
  const category = mapTypeToCategory(place.types || []);
  return {
    id: `gp_${place.place_id}`,
    place_id: place.place_id,
    name: place.name,
    description: null,
    address_line_1: place.vicinity || place.formatted_address || '',
    city: '',
    state: '',
    zip_code: '',
    latitude: place.geometry.location.lat,
    longitude: place.geometry.location.lng,
    average_rating: place.rating || 0,
    review_count: place.user_ratings_total || 0,
    price_level: place.price_level || null,
    category_name: category,
    categories: [{ id: 0, name: category, slug: category.toLowerCase().replace(/\s+/g, '-'), icon: null }],
    is_open_now: place.opening_hours?.open_now ?? null,
    photos: place.photos?.slice(0, 3).map(p => ({
      url: getPhotoUrl(p.photo_reference),
    })) || [],
    source: 'google_places' as const,
  };
};

export const formatPlaceDetailAsBusiness = (place: PlaceDetail) => {
  const category = mapTypeToCategory(place.types || []);

  let city = '';
  let state = '';
  let zipCode = '';
  if (place.address_components) {
    for (const comp of place.address_components) {
      if (comp.types.includes('locality')) city = comp.long_name;
      if (comp.types.includes('administrative_area_level_1')) state = comp.short_name;
      if (comp.types.includes('postal_code')) zipCode = comp.long_name;
    }
  }

  return {
    id: `gp_${place.place_id}`,
    place_id: place.place_id,
    name: place.name,
    description: null,
    address_line_1: place.formatted_address || '',
    city,
    state,
    zip_code: zipCode,
    phone: place.formatted_phone_number || null,
    website: place.website || null,
    latitude: place.geometry.location.lat,
    longitude: place.geometry.location.lng,
    average_rating: place.rating || 0,
    review_count: place.user_ratings_total || 0,
    price_level: place.price_level || null,
    category_name: category,
    categories: [{ id: 0, name: category, slug: category.toLowerCase().replace(/\s+/g, '-'), icon: null }],
    is_open_now: place.opening_hours?.open_now ?? null,
    hours: place.opening_hours?.weekday_text?.map((text, i) => ({
      day_of_week: (i + 1) % 7,
      text,
      is_closed: text.toLowerCase().includes('closed'),
    })) || [],
    photos: place.photos?.slice(0, 10).map(p => ({
      url: getPhotoUrl(p.photo_reference),
    })) || [],
    reviews: place.reviews?.map(r => ({
      id: `gpr_${r.time}`,
      rating: r.rating,
      content: r.text,
      createdAt: new Date(r.time * 1000).toISOString(),
      user: {
        username: r.author_name,
        firstName: r.author_name.split(' ')[0],
        lastName: r.author_name.split(' ').slice(1).join(' '),
        profileImageUrl: r.profile_photo_url || null,
      },
    })) || [],
    source: 'google_places' as const,
  };
};

File: backend_py/app/services/google_places.py
import os
import asyncio
from urllib.parse import quote
import httpx
from dotenv import load_dotenv

load_dotenv()

API_KEY = os.getenv("GOOGLE_MAPS_API_KEY", "")
BASE_URL = "https://places.googleapis.com/v1"

# --- Field masks ---

SEARCH_FIELDS = ",".join([
    "places.id", "places.displayName", "places.formattedAddress", "places.location",
    "places.rating", "places.userRatingCount", "places.priceLevel",
    "places.primaryType", "places.primaryTypeDisplayName",
    "places.regularOpeningHours.openNow", "places.photos", "places.businessStatus",
    # Used by chain_detector signals (website domain match, toll-free phone)
    "places.websiteUri", "places.internationalPhoneNumber",
    # Used by editorial signal — Google's natural-language summary, when present
    "places.editorialSummary",
])

DETAIL_FIELDS = ",".join([
    "id", "displayName", "formattedAddress", "location", "rating", "userRatingCount",
    "priceLevel", "primaryType", "primaryTypeDisplayName", "regularOpeningHours",
    "photos", "businessStatus", "internationalPhoneNumber", "websiteUri",
    "googleMapsUri", "reviews", "editorialSummary", "addressComponents",
])

# --- Category groups ---

CATEGORY_GROUPS = {
    "eatDrink": ["restaurant", "cafe", "bar", "bakery", "coffee_shop", "fast_food_restaurant"],
    "shopServices": ["shopping_mall", "clothing_store", "grocery_store", "hair_salon", "bank",
                     "book_store", "convenience_store", "electronics_store", "pet_store", "pharmacy"],
    "personalServices": ["gym", "spa", "beauty_salon", "car_repair", "car_wash", "laundry",
                         "veterinary_care", "dentist", "doctor"],
}

PRICE_LEVEL_MAP = {
    "PRICE_LEVEL_FREE": 0,
    "PRICE_LEVEL_INEXPENSIVE": 1,
    "PRICE_LEVEL_MODERATE": 2,
    "PRICE_LEVEL_EXPENSIVE": 3,
    "PRICE_LEVEL_VERY_EXPENSIVE": 4,
}


# --- Core API functions ---

async def search_nearby(latitude: float, longitude: float, radius: int = 1000,
                        included_types: list[str] | None = None, max_result_count: int = 10) -> list:
    if not API_KEY:
        return []
    try:
        body: dict = {
            "locationRestriction": {
                "circle": {
                    "center": {"latitude": latitude, "longitude": longitude},
                    "radius": min(radius, 50000),
                }
            },
            "maxResultCount": max_result_count,
        }
        if included_types:
            body["includedTypes"] = included_types

        async with httpx.AsyncClient() as client:
            resp = await client.post(
                f"{BASE_URL}/places:searchNearby",
                json=body,
                headers={
                    "X-Goog-Api-Key": API_KEY,
                    "X-Goog-FieldMask": SEARCH_FIELDS,
                    "Content-Type": "application/json",
                },
                timeout=10,
            )
            resp.raise_for_status()
            return resp.json().get("places", [])
    except Exception as e:
        print(f"Places API nearby search error: {e}")
        return []


async def search_nearby_diverse(latitude: float, longitude: float, radius: int = 1000) -> list:
    eat, shop, personal = await asyncio.gather(
        search_nearby(latitude, longitude, radius, CATEGORY_GROUPS["eatDrink"], 10),
        search_nearby(latitude, longitude, radius, CATEGORY_GROUPS["shopServices"], 10),
        search_nearby(latitude, longitude, radius, CATEGORY_GROUPS["personalServices"], 10),
    )

    groups = [eat, shop, personal]
    merged = []
    seen = set()
    max_len = max(len(g) for g in groups) if groups else 0

    for i in range(max_len):
        for group in groups:
            if i < len(group):
                place = group[i]
                pid = place.get("id")
                if pid and pid not in seen:
                    seen.add(pid)
                    merged.append(place)

    return merged[:20]


async def search_text(text_query: str, latitude: float, longitude: float, radius: int = 5000) -> list:
    if not API_KEY:
        return []
    try:
        async with httpx.AsyncClient() as client:
            resp = await client.post(
                f"{BASE_URL}/places:searchText",
                json={
                    "textQuery": text_query,
                    "locationBias": {
                        "circle": {
                            "center": {"latitude": latitude, "longitude": longitude},
                            "radius": min(radius, 50000),
                        }
                    },
                    "maxResultCount": 20,
                },
                headers={
                    "X-Goog-Api-Key": API_KEY,
                    "X-Goog-FieldMask": SEARCH_FIELDS,
                    "Content-Type": "application/json",
                },
                timeout=10,
            )
            resp.raise_for_status()
            return resp.json().get("places", [])
    except Exception as e:
        print(f"Places API text search error: {e}")
        return []


async def get_place_details(place_id: str):
    if not API_KEY:
        return None
    try:
        async with httpx.AsyncClient() as client:
            resp = await client.get(
                f"{BASE_URL}/places/{place_id}",
                headers={
                    "X-Goog-Api-Key": API_KEY,
                    "X-Goog-FieldMask": DETAIL_FIELDS,
                },
                timeout=10,
            )
            resp.raise_for_status()
            return resp.json()
    except Exception as e:
        print(f"Places API detail error: {e}")
        return None


def get_photo_url(photo_name: str, max_width: int = 400) -> str:
    return f"/api/photos?name={quote(photo_name)}&maxWidth={max_width}"


# --- Formatters ---

def format_place(place: dict) -> dict:
    price_str = place.get("priceLevel")
    price_level = PRICE_LEVEL_MAP.get(price_str) if price_str else None
    photos_raw = place.get("photos") or []
    photo_name = photos_raw[0].get("name") if photos_raw else None
    primary_type = place.get("primaryType")
    display_name_obj = place.get("primaryTypeDisplayName") or {}
    display_name = display_name_obj.get("text")

    return {
        "id": f"gp_{place.get('id', '')}",
        "place_id": place.get("id"),
        "name": (place.get("displayName") or {}).get("text", "Unknown"),
        "description": (place.get("editorialSummary") or {}).get("text"),
        "editorial_summary": (place.get("editorialSummary") or {}).get("text"),
        "address_line_1": place.get("formattedAddress", ""),
        "city": "",
        "state": "",
        "zip_code": "",
        "latitude": (place.get("location") or {}).get("latitude", 0),
        "longitude": (place.get("location") or {}).get("longitude", 0),
        "average_rating": place.get("rating", 0),
        "review_count": place.get("userRatingCount", 0),
        "price_level": price_level,
        "primary_type": primary_type,
        "primary_type_display_name": display_name,
        "category_name": display_name or (primary_type or "").replace("_", " ") or "Business",
        "types": [primary_type] if primary_type else [],
        "categories": [{"id": 0, "name": display_name, "slug": primary_type or "", "icon": None}] if display_name else [],
        "is_open_now": (place.get("regularOpeningHours") or {}).get("openNow"),
        "business_status": place.get("businessStatus"),
        # Surfaced from search results so chain_detector signals (website,
        # phone, editorial) can fire without a separate detail call.
        "website": place.get("websiteUri"),
        "phone": place.get("internationalPhoneNumber"),
        "photo_url": get_photo_url(photo_name, 800) if photo_name else None,
        "photos": [{"url": get_photo_url(p["name"], 800)} for p in photos_raw[:5] if "name" in p],
        "independence_score": None,
        "customer_facing_score": None,
        "local_badge": None,
        "source": "google_places",
    }


def format_place_detail(place: dict) -> dict:
    base = format_place(place)

    city = state = zip_code = ""
    for comp in place.get("addressComponents") or []:
        types = comp.get("types") or []
        if "locality" in types:
            city = comp.get("longText") or comp.get("shortText") or ""
        if "administrative_area_level_1" in types:
            state = comp.get("shortText") or ""
        if "postal_code" in types:
            zip_code = comp.get("longText") or ""

    weekday_descs = (place.get("regularOpeningHours") or {}).get("weekdayDescriptions") or []

    reviews_raw = place.get("reviews") or []
    reviews = []
    for r in reviews_raw:
        author = r.get("authorAttribution") or {}
        name_parts = (author.get("displayName") or "").split(" ", 1)
        reviews.append({
            "id": f"gpr_{r.get('publishTime', '')}",
            "rating": r.get("rating", 0),
            "title": None,
            "content": (r.get("text") or r.get("originalText") or {}).get("text", ""),
            "createdAt": r.get("publishTime", ""),
            "user": {
                "id": 0,
                "username": author.get("displayName", "Google User"),
                "firstName": name_parts[0] if name_parts else "",
                "lastName": name_parts[1] if len(name_parts) > 1 else "",
                "profileImageUrl": author.get("photoUri"),
            },
        })

    return {
        **base,
        "city": city,
        "state": state,
        "zip_code": zip_code,
        "phone": place.get("internationalPhoneNumber"),
        "website": place.get("websiteUri"),
        "google_maps_uri": place.get("googleMapsUri"),
        "weekday_descriptions": weekday_descs,
        "hours": [
            {"day_of_week": (i + 1) % 7, "text": text, "is_closed": "closed" in text.lower()}
            for i, text in enumerate(weekday_descs)
        ],
        "reviews": reviews,
    }
